docs/development/getting-started.rst

```diff
@@ -0,0 +1,96 @@
+===============
+Getting Started
+===============
+
+We’re pleased that you are interested in working on pip.
+
+This document is meant to get you setup to work on pip and to act as a guide and
+reference to the the development setup. If you face any issues during this
+process, please `open an issue`_ about it on the issue tracker.
+
+Development tools
+=================
+
+pip uses :pypi:`tox` for testing against multiple different Python environments
+and ensuring reproducible environments for linting and building documentation.
+
+For developing pip, you need to install ``tox`` on your system. Often, you can
+just do ``python -m pip install tox`` to install and use it.
+
+Running Tests
+-------------
+
+pip uses the :pypi:`pytest` test framework, :pypi:`mock` and :pypi:`pretend`
+for testing. These are automatically installed by tox for running the tests.
+
+To run tests locally, run:
+
+.. code-block:: console
+
+    $ tox -e py36
+
+The example above runs tests against Python 3.6. You can also use other
+versions like ``py27`` and ``pypy3``.
+
+pip's test suite requires supported version control tools (subversion, bazaar,
```

Comment:
Should we talk about using pytest-xdist (parallel testing) for faster test runs or link to pytest's CLI "advanced use" (things like markers, skipping network tests, running specific tests etc)?

Reply:
IMO, by far the most important thing to explain here is how to do the following:

1. Run the tests against whatever versions of Python you have installed locally. I imagine very few people have *all* the versions we test against in tox installed locally. Ideally, tox would cleanly skip Python versions that aren't installed, but from my recollection, it doesn't (it considers that as a failure).
2. Run specific tests, or a subset of tests. I personally always find it hard to remember how to do this, because it means remembering the specific pytest flags (which frankly I find a bit unintuitive) and how to specify them via tox.
3. Run coverage, linters, doc builds, things like that.

Those items are frankly "basic usage" for us - sure, parallel testing ans such things are advanced and maybe shouldn't be introduced here.

It's also worth remembering that the full test suite takes a *long* time to run, and expecting people to run it every time they make a small PR probably isn't reasonable - we have CI on PRs for that. I'd expect normal workflow to be

* Write a change, including a new test for it.
* Run the new test, against the unchanged code (to make sure it fails as expected) and the changed code (to make sure it succeeds with the change).
* *Maybe* run the full test suite. On Windows I almost never do that, as it's so slow.
* Push your PR and watch CI to see if anything gets picked up.

We should support that workflow explicitly, which means "run just one or a subset of the tests" is normal practice, not a special case.

Another note - some of the tox targets use basepython values like `python2` and `python3`. Those don't work on Windows. Either we need to fix `tox.ini` (which might involve some nasty hacks, or a fix in upstream tox to let such Python versions be specified in a platform-neutral way, the way `py36` works) or we need to document how to tweak `tox.ini` temporarily to run those targets (without ending up including the `tox.ini` change in your PR!). That's all a bit messy and I'd hoped to work on it at some point, but as I say, it may need changes to tox itself... Documenting the status quo is a good place to start, though.

Reply:
Let's fill in an issue to allow smart discovery of ``python3`` in tox; we'll add it something along the logic if ``python3`` is not found fallback to the corresponding available ``python3.7``, ``python3.6``, ``python3.5`` etc. 👍 I consider this a valid improvement request, and will likely be added.

Reply:
As for running a given test with tox one would need to do ``tox -e py27 -- -k my_test_name``. Tox itself skips missing interpreters, see https://github.com/tox-dev/tox/blob/master/tox.ini#L11 .

Reply:
Either way let me know what else would you want from tox. We can help.

Reply:
That is a separate enhancement that I'd welcome.

IIUC, on Windows, something like a python2.7 won't work since the executables there are all python.exe for various python versions. The canonical way to run and manage different versions of python is using `py`, which works out of the box (mostly). I think a solution to what @pfmoore is saying is that tox should try using py -2.7 on Windows (and then fallback the full name for backward compat; in whichever order) so that things just work out of the box on Windows.

Reply:
``py -x.y``?

Reply:
👍 sure, I'm available with limited amount during the weekend but let's pick it up next week

Reply:
@gaborbernat Thanks for picking up on this. IIRC, I actually [contributed](https://github.com/tox-dev/tox/commit/ab59af325080701ac12e8484d8bb120e503be888) to the code for tox that interprets the `py36` etc standard names on Windows, and at the time I missed that it didn't cover `python3` and `python2`. I'm happy to help if you have any questions - although I'm not around tomorrow and have limited time over the next week or two.

Reply:
Nope, but `py36` would work cross-platform, AIUI.